### ISSUE TITLE
Cannot access a disposed object.\nObject name: 'System.Net.Sockets.Socket

### ISSUE THREAD

**Issue by PaulPBastian** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1587]
Hi Team, I am using  EF core  3.1.19 to connect to AWS RDS MYSQL database.  The packages used are below .
```
   <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.19" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.19" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="3.1.19" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="3.1.19" />
    <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="3.2.7" />
```
and ```MYSQL Connector``` used is 0.69.10 version 

I am getting the below exception intermittently while doing some db operation . The inner exception states that Cannot access a disposed object.\nObject name: 'System.Net.Sockets.Socket

 I recently migrated from EF core 2.7 which was working fine there.

### Exception Details

```
Message": "Connect Timeout expired.",
  "Source": "MySqlConnector",
  "StackTraceString": "   at MySqlConnector.Core.ServerSession.OpenTcpSocketAsync(ConnectionSettings cs, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken)\n   at MySqlConnector.Core.ServerSession.ConnectAsync(ConnectionSettings cs, Int32 startTickCount, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 368\n   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 143\n   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 143\n   at MySql.Data.MySqlClient.MySqlConnection.CreateSessionAsync(ConnectionPool pool, Int32 startTickCount, Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlConnection.cs:line 758",
  "RemoteStackTraceString": null,
  "RemoteStackIndex": 0,
  "HResult": -2147467259,
  "HelpURL": null
},
{
  "Depth": 4,
  "ClassName": "System.ObjectDisposedException",
  "Message": "Cannot access a disposed object.\nObject name: 'System.Net.Sockets.Socket'.",
  "Source": "System.Net.Sockets",
  "StackTraceString": "   at System.Net.Sockets.Socket.BeginConnect(IPAddress address, Int32 port, AsyncCallback requestCallback, Object state)\n   at System.Net.Sockets.TcpClient.BeginConnect(IPAddress address, Int32 port, AsyncCallback requestCallback, Object state)\n   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2](Func`5 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, Object state, TaskCreationOptions creationOptions)\n   at System.Net.Sockets.TcpClient.ConnectAsync(IPAddress address, Int32 port)\n   at MySqlConnector.Core.ServerSession.OpenTcpSocketAsync(ConnectionSettings cs, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 907",
```

### Further technical details

MySQL version: 5.7.33
Operating system: Linux 64 AWS ECS
Pomelo.EntityFrameworkCore.MySql version:3.2.7
Microsoft.AspNetCore.App vesion:  3.1

Other details about my project setup:
connection string
Server=serverurl;Port=3306;Database=DBName;User=******;Password=********;Allow User Variables=True;SslMode=Required

i am registering the DB context as InstancePerLifetimeScope.

 

Please help me on this.

**Comment by lauxjpn** [comment]
Maybe this is this one:

_From [Version History: 2.0.0](https://mysqlconnector.net/overview/version-history/#2-0-0):_
> Fix rare `ObjectDisposedException` that could be thrown when a connection timeout occurred.

If so, there is no fix available that is compatible with Pomelo 3.x.
Since you didn't post the full stack trace, it is hard to say how to fix this in a simple fashion. (The general way to deal with it would probably be to retry the operation.)

You could also upgrade Pomelo to a newer version.

**Comment by bgrainger** [comment]
The exception message in the bug title is spurious: that's the inner exception. The real issue is `MySqlException: Connect Timeout expired.`

Without additional information, it looks like a network timeout is occurring when connecting; that's the problem that needs to be diagnosed further.

(And FWIW, no fixes are planned for the v0.x (nor probably the v1.x) series of MySqlConnector, so I would also echo the advice to upgrade.)

**Comment by PaulPBastian** [comment]
Hi @lauxjpn/ @bgrainger  ,

Thanks for your input. 

More details and some workarounds done from my side. 

Application include multiple microservices running in separate thread and the above exception occurred for a service having some db operation running parallely as below(sample code)
```
            foreach (var id in idCollection)
            {
                var updateTask = RunCommandInParallelundersinglelifetimescope(async m => someDBUpdateOperation(id));
                updateTasks.Add(updateTask);
            }

            await Task.WhenAll(updateTasks.ToArray());

```
This code was working fine in previous version(2.7.1 EF and its related dependencies) . 
Now as part of above issue, i have tried removing WhenAll() and make it as a separate task with separate life time scope . 

Now the issue is gone, but wanted to know the details why it happened. even if it happens , it should be a command timeout and not a connection time out. right? 
or am i missing some connection timeout explicitly to be mentioned?

```

  {
    "Depth": 2,
    "ClassName": "MySql.Data.MySqlClient.MySqlException",
    "Message": "Connect Timeout expired.",
    "Source": "MySqlConnector",
    "StackTraceString": "   at MySql.Data.MySqlClient.MySqlConnection.CreateSessionAsync(ConnectionPool pool, Int32 startTickCount, Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlConnection.cs:line 758\n   at MySql.Data.MySqlClient.MySqlConnection.OpenAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlConnection.cs:line 416\n   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnectionAsync(Boolean errorsExpected, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnectionAsync(Boolean errorsExpected, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenAsync(CancellationToken cancellationToken, Boolean errorsExpected)\n   at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlRelationalConnection.OpenAsync(CancellationToken cancellationToken, Boolean errorsExpected)\n   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.AsyncEnumerator.InitializeReaderAsync(DbContext _, Boolean result, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "HResult": -2147467259,
    "HelpURL": null
  },
  {
    "Depth": 3,
    "ClassName": "MySql.Data.MySqlClient.MySqlException",
    "Message": "Connect Timeout expired.",
    "Source": "MySqlConnector",
    "StackTraceString": "   at MySqlConnector.Core.ServerSession.OpenTcpSocketAsync(ConnectionSettings cs, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken)\n   at MySqlConnector.Core.ServerSession.ConnectAsync(ConnectionSettings cs, Int32 startTickCount, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 368\n   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 143\n   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 143\n   at MySql.Data.MySqlClient.MySqlConnection.CreateSessionAsync(ConnectionPool pool, Int32 startTickCount, Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlConnection.cs:line 758",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "HResult": -2147467259,
    "HelpURL": null
  },
  {
    "Depth": 4,
    "ClassName": "System.ObjectDisposedException",
    "Message": "Cannot access a disposed object.\nObject name: 'System.Net.Sockets.Socket'.",
    "Source": "System.Net.Sockets",
    "StackTraceString": "   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)\n   at System.Net.Sockets.TcpClient.EndConnect(IAsyncResult asyncResult)\n   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)\n--- End of stack trace from previous location where exception was thrown ---\n   at MySqlConnector.Core.ServerSession.OpenTcpSocketAsync(ConnectionSettings cs, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 907",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "HResult": -2146232798,
    "HelpURL": null
  }
  ```

Also for some other service, command is also timing out intermittently(like once an hr) with below details(not sure if i can add along with this issue, but may be helpful to you.)

```
 {
    "Depth": 2,
    "ClassName": "MySql.Data.MySqlClient.MySqlException",
    "Message": "The Command Timeout expired before the operation completed.",
    "Source": "MySqlConnector",
    "StackTraceString": "   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet() in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlDataReader.cs:line 116\n   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlDataReader.cs:line 391\n   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 62\n   at MySql.Data.MySqlClient.MySqlCommand.ExecuteDbDataReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlCommand.cs:line 257\n   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.AsyncEnumerator.InitializeReaderAsync(DbContext _, Boolean result, CancellationToken cancellationToken)\n   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "HResult": -2147467259,
    "HelpURL": null
  },
  {
    "Depth": 3,
    "ClassName": "MySql.Data.MySqlClient.MySqlException",
    "Message": "The Command Timeout expired before the operation completed.",
    "Source": "MySqlConnector",
    "StackTraceString": "   at MySqlConnector.Protocol.Serialization.StreamByteHandler.<ReadBytesAsync>g__DoReadBytesAsync|6_2(Memory`1 buffer_) in /_/src/MySqlConnector/Protocol/Serialization/StreamByteHandler.cs:line 91\n   at MySqlConnector.Protocol.Serialization.BufferedByteReader.ReadBytesAsync(IByteHandler byteHandler, ArraySegment`1 buffer, Int32 totalBytesToRead, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/BufferedByteReader.cs:line 36\n   at MySqlConnector.Protocol.Serialization.ProtocolUtility.<ReadPacketAsync>g__AddContinuation|1_0(ValueTask`1 headerBytes_, BufferedByteReader bufferedByteReader_, IByteHandler byteHandler_, Func`1 getNextSequenceNumber_, ProtocolErrorBehavior protocolErrorBehavior_, IOBehavior ioBehavior_) in /_/src/MySqlConnector/Protocol/Serialization/ProtocolUtility.cs:line 413\n   at MySqlConnector.Protocol.Serialization.ProtocolUtility.<DoReadPayloadAsync>g__AddContinuation|5_0(ValueTask`1 readPacketTask_, BufferedByteReader bufferedByteReader_, IByteHandler byteHandler_, Func`1 getNextSequenceNumber_, ArraySegmentHolder`1 previousPayloads_, ProtocolErrorBehavior protocolErrorBehavior_, IOBehavior ioBehavior_) in /_/src/MySqlConnector/Protocol/Serialization/ProtocolUtility.cs:line 484\n   at MySqlConnector.Core.ServerSession.ReceiveReplyAsyncAwaited(ValueTask`1 task) in /_/src/MySqlConnector/Core/ServerSession.cs:line 811\n   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 51",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "HResult": -2147467259,
    "HelpURL": null
  },
  {
    "Depth": 4,
    "ClassName": "System.IO.IOException",
    "Message": "Unable to read data from the transport connection: Operation canceled.",
    "Source": "System.Net.Sockets",
    "StackTraceString": "   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)\n   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)\n   at System.Net.Security.SslStream.<FillBufferAsync>g__InternalFillBufferAsync|215_0[TReadAdapter](TReadAdapter adap, ValueTask`1 task, Int32 min, Int32 initial)\n   at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)\n   at MySqlConnector.Protocol.Serialization.StreamByteHandler.<ReadBytesAsync>g__DoReadBytesAsync|6_2(Memory`1 buffer_) in /_/src/MySqlConnector/Protocol/Serialization/StreamByteHandler.cs:line 91",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "HResult": -2146232800,
    "HelpURL": null
  },
  {
    "Depth": 5,
    "ClassName": "System.Net.Sockets.SocketException",
    "Message": "Operation canceled",
    "Source": null,
    "StackTraceString": null,
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "HResult": -2147467259,
    "HelpURL": null
  }
  ```

there is some constraints at this point to upgrade Ef core 5.X or higher, so looking for a workaround at this point.

Thanks in Advance.

**Comment by bgrainger** [comment]
> Now the issue is gone, but wanted to know the details why it happened. even if it happens , it should be a command timeout and not a connection time out. right?

Not necessarily. A "command timeout" occurs when `MySqlCommand.ExecuteX` takes longer than `.CommandTimeout` seconds to execute. A "connection timeout" occurs when `MySqlConnection.Open` takes longer than `MySqlConnectionStringBuilder.ConnectTimeout` seconds to open. This can happen if you have more concurrent requests to `Open` than you have available connections in your connection pool. In that scenario, if it takes longer than 15 seconds (the default) to process the backlog of connections, you will get a timeout.

You could increase `MaxPoolSize` but probably the better fix is reducing the number of concurrent database operations you're attempting to perform.

**Comment by lauxjpn** [comment]
@PaulPBastian Was each of your in parallel running threads using their own `DbContext` instance? Because EF Core does not support using the same context in parallel (though you _can_ using the same context on multiple threads, as long as only one thread calls into the `DbContext` at the same time).

Also, in case the app is running as an ASP.NET Core app, if the tasks are using any scoped services, the request cannot return before all tasks are done using the services, or those services might get disposed before your tasks are done.

To check whether this is a connection pool/resource issue (as pointed out by @bgrainger), you should check the number of open connections from that app instance on that database instance.

**Comment by bgrainger** [comment]
FWIW the Inner Exception was removed in https://github.com/mysql-net/MySqlConnector/issues/1035 (>= 1.3.13) to avoid the confusing error message in the OP.

**Comment by PaulPBastian** [comment]
Was each of your in parallel running threads using their own DbContext instance? 
Yes @lauxjpn , Each concurrent jobs will be running on separate thread and all db operation will happen under a lifetimescope instance created for each job like below
```
  using (var childScope = _scope.BeginLifetimeScope())
   { 
     _logger.Debug("Executing job");
    }
 ```

and dbcontext is resistered as 

builder.Register({dbconext}).InstancePerLifetimeScope();

let me check the number of open connections

**Comment by lauxjpn** [comment]
> let me check the number of open connections

@PaulPBastian Were you able to trace them?
